### PR TITLE
(maint) Run service provider tests on appropriate platforms

### PR DESCRIPTION
**Pull request by joshcooper**
Also updates upstart provider tests to not reload the provider

### CODE REVIEW

File: spec/unit/provider/service/base_spec.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 require 'rbconfig'
 require 'fileutils'
 
-describe "base service provider" do
+describe "base service provider", unless: Puppet::Util::Platform.windows? || Puppet::Util::Platform.jruby? do
```
Comment (kris-bosland) [inline spec/unit/provider/service/base_spec.rb]: I recognize that we are not using these providers in our server, but are other users using JRuby?  Do we support Puppet as a gem running on JRuby?

File: spec/unit/provider/service/bsd_spec.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 require 'spec_helper'
 
 describe 'Puppet::Type::Service::Provider::Bsd',
-         :unless => Puppet::Util::Platform.windows? || Puppet::Util::Platform.jruby? do
+         unless: Puppet::Util::Platform.windows? || Puppet::Util::Platform.jruby? do
```
Comment (kris-bosland) [inline spec/unit/provider/service/bsd_spec.rb]: Did this not work before, or is this for readability/uniformity?

File: spec/unit/provider/service/openbsd_spec.rb [hunk 3]
```diff
@@ -191,7 +191,7 @@
   end
 
   context "#flags=" do
-    it "should run rcctl to set flags", unless: Puppet::Util::Platform.windows? || RUBY_PLATFORM == 'java' do
+    it "should run rcctl to set flags" do
```
Comment (kris-bosland) [inline spec/unit/provider/service/openbsd_spec.rb]: Why is this one not excluded from windows anymore?

File: spec/unit/provider/service/windows_spec.rb [hunk 4]
```diff
@@ -1,8 +1,7 @@
 require 'spec_helper'
 require 'win32/service' if Puppet::Util::Platform.windows?
 
-describe 'Puppet::Type::Service::Provider::Windows',
-    :if => Puppet::Util::Platform.windows? && !Puppet::Util::Platform.jruby? do
+describe 'Puppet::Type::Service::Provider::Windows', if: Puppet::Util::Platform.windows? do
```
Comment (kris-bosland) [inline spec/unit/provider/service/windows_spec.rb]: Are we allowing JRuby for running the windows service tests?

**Comment by branan** [comment]
This feels like it's the opposite direction of where we should be going, to me. Instead of restricting more unit tests to specific platforms, we should figure out where our abstractions and mocks are leaky and fix them. A developer should be able to trust `rake spec` on their laptop to give them a *reasonable* picture of whether code they've touched has broken other parts of the product. The more we treat unit tests as integration tests and let the underlying system leak into them, the less coverage and confidence we have when making changes to the codebase.

**Comment by joshcooper** [comment]
Yeah that's fair. 

> we should figure out where our abstractions and mocks are leaky and fix them

Yeah the current specs workaround $CHILD_STATUS on Windows, which shows that the current execution API does not do a good job of doing that. Later on Andy modified the API to return a ProcessStatus object instead of just the output.

I'll close this out and just resubmit just the upstart spec changes.

**Pull request closed without merging**